### ISSUE TITLE
Add pre-built dictionary of "did you mean..." suggestions

### ISSUE THREAD

**Issue by schlessera** [wp-cli/wp-cli#4379]
For https://github.com/wp-cli/wp-cli/issues/3861 we built an algorithm that compares an invalid parameter to all possible parameters based on Levenshtein distance and makes suggestions if the Levenshtein distance is below or equal a set threshold (currently 2).

We still need to add a dictionary of pre-built "invalid parameter" => "suggestion" mappings so that we can provide meaningful suggestions that more drastically deviate from what was typed, without the need to bump the threshold higher. A too high threshold will show more meaningless suggestions.

The dictionary should just be an associative array that takes precedence over the calculated suggestion:

```php
function get_suggestion( $target, array $options, $threshold = 2 ) {
   $suggestion_map = array(
      'dbpassword' => 'dbpass',
      // [...]
   );

   if ( array_key_exists( $target, $suggestion_map ) ) {
      return $suggestion_map[ $target ];
   }

   // [...]
}
```

**Comment by szepeviktor** [comment]
How do git and drush do it?
Something like http://php.net/manual/en/function.levenshtein.php ?

**Comment by szepeviktor** [comment]
Git by Levenshtein:
https://github.com/git/git/blob/e0aaa1b6532cfce93d87af9bc813fb2e7a7ce9d7/help.c#L302

**Comment by szepeviktor** [comment]
Drush does not try to correct typos:
```
$ drush watchdog-shoe
The drush command 'watchdog-shoe' could not be found. 
```

**Comment by szepeviktor** [comment]
👍 for Levenshtein.

**Comment by schlessera** [comment]
@szepeviktor We are already using `levenshtein()`: https://github.com/wp-cli/wp-cli/blob/c79ff2e6895a468b9783cc7bafe1d6c8f4ca6895/php/utils.php#L1077

The issue above is for the cases where the distance would be so large that we cannot provide a meaningful threshold anymore. A high threshold just produces random suggestions.

**Comment by szepeviktor** [comment]
Thank you.
Now-a-days I am not up-to-date with WP-CLI.

**Comment by schlessera** [comment]
Ha, no worries. See it as a confirmation that we all found your idea worthwhile! ;)

**Comment by danielbachhuber** [comment]
> We still need to add a dictionary of pre-built "invalid parameter" => "suggestion" mappings so that we can provide meaningful suggestions that more drastically deviate from what was typed, without the need to bump the threshold higher. A too high threshold will show more meaningless suggestions.

How should we produce this dictionary?

**Comment by schlessera** [comment]
Right now, we have 1 entry:
```
'dbpassword' => 'dbpass'
```
I'd suggest starting with that one and adding whenever we find a type/wrong parameter name that we think should produce a meaningful suggestion but doesn't.

**Comment by danielbachhuber** [comment]
> I'd suggest starting with that one and adding whenever we find a type/wrong parameter name that we think should produce a meaningful suggestion but doesn't.

Sounds good. To address this issue fully, it'd be great to have someone spend ~30 minutes reviewing arguments for all bundled WP-CLI commands and produce a list of a dozen mapping suggestions we can start with.

**Comment by n8finch** [comment]
> Sounds good. To address this issue fully, it'd be great to have someone spend ~30 minutes reviewing arguments for all bundled WP-CLI commands and produce a list of a dozen mapping suggestions we can start with.

Just to clarify, are you looking for parameters only, or subcommands as well? 

For example:

- Subcommand should be: `wp cli check-update`
- Array could include: `'check' => 'check-update' , 'check update' => 'check-update'`
- Errors received:
```
Error: 'check' is not a registered subcommand of 'cli'. See 'wp help cli' for available subcommands.
```

Or

- Command with parameters should be: `wp core download --locale=en --version=4.2`
- Array could include: `'language' => 'local' , 'v' => 'version'`
Error received: 
```
Error: Parameter errors: unknown --language parameter
Error: Parameter errors: unknown --v parameter
```

If those are the droids--I mean arrays--you're looking for, I'd be happy to take this on and get the associative array rolling. 🤓

**Comment by danielbachhuber** [comment]
> Just to clarify, are you looking for parameters only, or subcommands as well?

Your latter example for now — command parameters.

**Comment by n8finch** [comment]
Great! I'll get on that.

**Comment by schlessera** [comment]
Fixed with #4392
